snuba_sdk/query_validation.py

```diff
@@ -0,0 +1,224 @@
+from __future__ import annotations
+
+from typing import List, Optional, Sequence, Set, Tuple, Union
+
+# Import the modules due to sphinx autodoc problems
+# https://github.com/agronholm/sphinx-autodoc-typehints#dealing-with-circular-imports
+from snuba_sdk import query as main
+from snuba_sdk import query_visitors as qvisitors
+from snuba_sdk.column import Column
+from snuba_sdk.conditions import (
+    BooleanCondition,
+    Condition,
+    Op,
+    get_top_level_conditions,
+)
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import Expression
+from snuba_sdk.function import CurriedFunction
+from snuba_sdk.relationships import Join
+from snuba_sdk.schema import Column as ColumnModel
+
+
+class InvalidMatchError(Exception):
+    pass
+
+
+def validate_match(
+    query: main.Query, column_finder: qvisitors.ExpressionSearcher
+) -> None:
+    """
+    Perform validation related to the match clause of the query. Currently that takes two forms:
+    validate that the columns referenced in the query are valid columns based on the match clause,
+    and ensure that the query has the appropriate conditions based on the entities required columns.
+
+    :param query: The query to be validated.
+    :type query: main.Query
+    :param column_finder: An ExpressionSearcher used to find all the columns in the query.
+    :type column_finder: ExpressionSearcher
+    """
+    all_columns = column_finder.visit(query)
+    if isinstance(query.match, main.Query):
+        validate_subquery(query.match, all_columns)
+        validate_match(query.match, column_finder)
+    elif isinstance(query.match, Join):
+        validate_join(query.match, all_columns)
+    else:
+        validate_entity(query.match, all_columns)
+
+    validate_required_columns(query)
+
+
+def validate_subquery(match: main.Query, all_columns: Set[Expression]) -> None:
+    """
+    Validate that the outer query is only referencing columns in the inner query.
+
+    :param match: The inner query of the query being validated.
+    :type match: main.Query
+    :param all_columns: All the columns referenced in the outer query.
+    :type all_columns: Set[Expression]
+    :raises InvalidQueryError
+    """
+    inner_match = set()
+    assert match.select is not None
+    for s in match.select:
+        if isinstance(s, CurriedFunction):
+            inner_match.add(s.alias)
+        elif isinstance(s, Column):
+            inner_match.add(s.name)
+
+    for c in all_columns:
+        if isinstance(c, Column) and c.name not in inner_match:
+            raise InvalidMatchError(
+                f"outer query is referencing column '{c.name}' that does not exist in subquery"
+            )
+
+
+def validate_join(match: Join, all_columns: Set[Expression]) -> None:
+    """
+    Validate that all the columns in the query are referencing an entity
+    in the match, and that there is no alias shadowing.
```

Comment:
Giving two different entities the same alias.

Reply:
why not just say that? It's more clear

Reply:
Because shadowing is an established technical concept.

Reply:
I don't think it's so well established. If you google alias shadowing(2), only by the sixth result do you get a definition mixed with other completely unrelated results. It's also a well established practice of technical writing to use plain language as often as possible(1). The difference between 

`that there is no no alias shadowing`
`that no different entities have the same alias` 

is 12 characters, but one is much clearer to more people.

I understand that I appear pedantic for this, but it's not this term specifically that I have issue with. If you feel strongly about it you can leave it as is. The reason I'm pressing on these issues is because we want to be the enablers of features rather than the gatekeepers of features. To do that, what we create and the intentions behind it has to be very clear to others in the organization. That means our code and our documentation has to be accessible. When other teams or members of the open source community can easily extend what we do because our thought process is clear from what we've done, we create less work for ourselves by making the code and documentation we write an effective medium of communication.


[1] https://www.plainlanguage.gov/guidelines/words/avoid-jargon/
[2] https://www.google.com/search?hl=en&q=alias%20shadowing